dialer.go

```diff
@@ -225,6 +262,119 @@ func (d *Dialer) Dial(ctx context.Context, instance string, opts ...DialOption)
 	}), nil
 }
 
+// metadataExchange sends metadata about the connection prior to the database
+// protocol taking over. The exchange consists of four steps:
+//
+//  1. Prepare a MetadataExchangeRequest including the IAM Principal's OAuth2
+//     token, the user agent, and the requested authentication type.
+//
+//  2. Write the size of the message as a big endian uint32 (4 bytes) to the
+//     server followed by the marshaled message. The length does not include the
+//     initial four bytes.
+//
+//  3. Read a big endian uint32 (4 bytes) from the server. This is the
+//     MetadataExchangeResponse message length and does not include the initial
+//     four bytes.
+//
+//  4. Unmarshal the response using the message length in step 3. If the
+//     response is not OK, return the response's error. If there is no error, the
+//     metadata exchange has succeeded and the connection is complete.
+//
+// Subsequent interactions with the test server use the database protocol.
+func (d *Dialer) metadataExchange(conn net.Conn) error {
+	tok, err := d.tokenSource.Token()
+	if err != nil {
+		return err
+	}
+	authType := connectorspb.MetadataExchangeRequest_DB_NATIVE
+	if d.useIAMAuthN {
+		authType = connectorspb.MetadataExchangeRequest_AUTO_IAM
+	}
+	req := &connectorspb.MetadataExchangeRequest{
+		UserAgent:   d.userAgent,
+		AuthType:    authType,
+		Oauth2Token: tok.AccessToken,
+	}
+	m, err := proto.Marshal(req)
+	if err != nil {
+		return err
+	}
+	b := d.buffer.get()
+	defer d.buffer.put(b)
+
+	buf := *b
+	reqSize := proto.Size(req)
+	binary.BigEndian.PutUint32(buf, uint32(reqSize))
+	buf = append(buf[:4], m...)
+
+	// Set IO deadline before write
+	err = conn.SetDeadline(time.Now().Add(ioTimeout))
+	if err != nil {
+		return err
+	}
+	defer conn.SetDeadline(time.Time{})
+
+	_, err = conn.Write(buf)
+	if err != nil {
+		return err
+	}
+
+	// Reset IO deadline before read
+	err = conn.SetDeadline(time.Now().Add(ioTimeout))
```

Comment:
You have to clear the deadline too similar to line #315?

Reply:
Yes, good catch. Fixed.